File: src/memory/search/vector.ts
/**
 * Vector search with cosine similarity
 *
 * Embeds query text and ranks lessons by semantic similarity.
 * Uses SQLite cache to avoid recomputing embeddings.
 */

import { readCctPatterns, type CctPattern } from '../../compound/index.js';
import { embedText } from '../embeddings/index.js';
import { isModelAvailable } from '../embeddings/model-info.js';
import { contentHash, getCachedEmbeddingsBulk, getCachedInsightEmbedding, readAllFromSqlite, setCachedEmbedding, setCachedInsightEmbedding, syncIfNeeded } from '../storage/index.js';
import type { MemoryItem } from '../types.js';

/**
 * In-memory embedding cache for CCT patterns.
 * CCT patterns don't have rows in the SQLite lessons table,
 * so setCachedEmbedding (UPDATE-only) is a no-op for them.
 * This Map caches embeddings keyed by "id:contentHash".
 *
 * NOTE: This cache is intentionally unbounded. Compound Agent is a CLI tool
 * (one process per command), so entries are bounded by a single command's
 * lifetime. Do not use this pattern in long-running servers.
 */
const cctEmbeddingCache = new Map<string, Float32Array>();

/** Clear the CCT embedding cache. Exported for testing. */
export function clearCctEmbeddingCache(): void {
  cctEmbeddingCache.clear();
}

/**
 * Calculate cosine similarity between two vectors.
 * Returns value between -1 (opposite) and 1 (identical).
 */
export function cosineSimilarity(a: ArrayLike<number>, b: ArrayLike<number>): number {
  if (a.length !== b.length) {
    throw new Error('Vectors must have same length');
  }

  let dotProduct = 0;
  let normA = 0;
  let normB = 0;

  for (let i = 0; i < a.length; i++) {
    dotProduct += a[i]! * b[i]!;
    normA += a[i]! * a[i]!;
    normB += b[i]! * b[i]!;
  }

  const magnitude = Math.sqrt(normA) * Math.sqrt(normB);
  if (magnitude === 0) return 0;

  return dotProduct / magnitude;
}

/**
 * Memory item with similarity score.
 * The `lesson` field holds any MemoryItem type (not just Lesson).
 * Field name kept for backward compatibility.
 */
export interface ScoredLesson {
  lesson: MemoryItem;
  score: number;
}

/** Options for vector search */
export interface SearchVectorOptions {
  /** Maximum number of results to return (default: 10) */
  limit?: number;
}

/** Default number of results to return */
const DEFAULT_LIMIT = 10;

/**
 * Search lessons by vector similarity to query text.
 * Returns top N lessons sorted by similarity score (descending).
 * Uses embedding cache to avoid recomputing embeddings.
 */
/**
 * Convert a CctPattern to a MemoryItem-like shape for search results.
 */
function cctToMemoryItem(pattern: CctPattern): MemoryItem {
  return {
    id: pattern.id,
    type: 'lesson',
    trigger: pattern.name,
    insight: pattern.description,
    tags: [],
    source: 'manual',
    context: { tool: 'compound', intent: 'synthesis' },
    created: pattern.created,
    confirmed: true,
    supersedes: [],
    related: pattern.sourceIds,
  };
}

export async function searchVector(
  repoRoot: string,
  query: string,
  options?: SearchVectorOptions
): Promise<ScoredLesson[]> {
  const limit = options?.limit ?? DEFAULT_LIMIT;
  // Ensure SQLite cache is fresh, then read from it (avoids redundant JSONL parse)
  await syncIfNeeded(repoRoot);
  const items = readAllFromSqlite(repoRoot);

  // Read CCT patterns if available
  let cctPatterns: CctPattern[] = [];
  try {
    cctPatterns = await readCctPatterns(repoRoot);
  } catch {
    // File doesn't exist or is unreadable — proceed without CCT patterns
  }

  if (items.length === 0 && cctPatterns.length === 0) return [];

  // Embed the query
  const queryVector = await embedText(query);

  // Bulk-read all cached embeddings in one query (instead of N individual reads)
  const cachedEmbeddings = getCachedEmbeddingsBulk(repoRoot);

  // Score each item, skipping invalidated ones
  const scored: ScoredLesson[] = [];
  for (const item of items) {
    // Skip invalidated items
    if (item.invalidatedAt) continue;

    try {
      const itemText = `${item.trigger} ${item.insight}`;
      const hash = contentHash(item.trigger, item.insight);

      // Try bulk cache first
      const cached = cachedEmbeddings.get(item.id);
      let itemVector: Float32Array;

      if (cached && cached.hash === hash) {
        itemVector = cached.vector;
      } else {
        // Cache miss or stale - compute and store
        itemVector = await embedText(itemText);
        setCachedEmbedding(repoRoot, item.id, itemVector, hash);
      }

      const score = cosineSimilarity(queryVector, itemVector);
      scored.push({ lesson: item, score });
    } catch {
      // Skip items that fail embedding — return partial results
      continue;
    }
  }

  // Score CCT patterns (use in-memory cache since they lack SQLite rows)
  for (const pattern of cctPatterns) {
    try {
      const text = `${pattern.name} ${pattern.description}`;
      const hash = contentHash(pattern.name, pattern.description);
      const cacheKey = `${pattern.id}:${hash}`;

      let vec = cctEmbeddingCache.get(cacheKey);
      if (!vec) {
        vec = await embedText(text);
        cctEmbeddingCache.set(cacheKey, vec);
      }

      const score = cosineSimilarity(queryVector, vec);
      scored.push({ lesson: cctToMemoryItem(pattern), score });
    } catch {
      continue;
    }
  }

  // Sort by score descending and take top N
  scored.sort((a, b) => b.score - a.score);
  return scored.slice(0, limit);
}

export interface SimilarLesson {
  item: MemoryItem;
  score: number;
}

export interface FindSimilarOptions {
  threshold?: number;
  excludeId?: string;
  /** Pre-loaded items to search. When provided, skips readMemoryItems(). */
  items?: MemoryItem[];
}

const DEFAULT_THRESHOLD = 0.80;

/**
 * Find lessons semantically similar to the given text.
 * Embeds using insight text only (not trigger) to avoid noise from generic triggers.
 * Does NOT include CCT patterns.
 */
export async function findSimilarLessons(
  repoRoot: string,
  text: string,
  options?: FindSimilarOptions
): Promise<SimilarLesson[]> {
  const threshold = options?.threshold ?? DEFAULT_THRESHOLD;
  const excludeId = options?.excludeId;

  if (!isModelAvailable()) return [];

  let items: MemoryItem[];
  if (options?.items) {
    items = options.items;
  } else {
    await syncIfNeeded(repoRoot);
    items = readAllFromSqlite(repoRoot);
  }
  if (items.length === 0) return [];

  const queryVector = await embedText(text);

  const scored: SimilarLesson[] = [];
  for (const item of items) {
    if (item.invalidatedAt) continue;
    if (excludeId && item.id === excludeId) continue;

    try {
      // Use insight ONLY for embedding (NOT trigger + insight).
      // Stored in separate columns to avoid cache conflicts with searchVector.
      const hash = contentHash(item.insight, '');
      let itemVector = getCachedInsightEmbedding(repoRoot, item.id, hash);

      if (!itemVector) {
        itemVector = await embedText(item.insight);
        setCachedInsightEmbedding(repoRoot, item.id, itemVector, hash);
      }

      const score = cosineSimilarity(queryVector, itemVector);
      if (score >= threshold) {
        scored.push({ item, score });
      }
    } catch {
      continue;
    }
  }

  scored.sort((a, b) => b.score - a.score);
  return scored;
}

File: go/internal/search/vector.go
package search

import (
	"database/sql"
	"sort"

	"github.com/nathandelacretaz/compound-agent/internal/compound"
	"github.com/nathandelacretaz/compound-agent/internal/memory"
	"github.com/nathandelacretaz/compound-agent/internal/storage"
	"github.com/nathandelacretaz/compound-agent/internal/util"
)

// DefaultSimilarityThreshold is the default cosine similarity threshold
// for FindSimilarLessons.
const DefaultSimilarityThreshold = 0.80

// maxEmbedBatch is the maximum texts per Embed() call, matching the Rust
// daemon's CA_EMBED_MAX_BATCH default. Larger batches are chunked automatically.
const maxEmbedBatch = 64

// Embedder provides text embedding functionality.
// Implemented by the embed daemon client.
type Embedder interface {
	Embed(texts []string) ([][]float64, error)
}

// embedBatched embeds texts in chunks of maxEmbedBatch to stay within daemon limits.
func embedBatched(embedder Embedder, texts []string) ([][]float64, error) {
	if len(texts) == 0 {
		return nil, nil
	}
	result := make([][]float64, 0, len(texts))
	for i := 0; i < len(texts); i += maxEmbedBatch {
		end := i + maxEmbedBatch
		if end > len(texts) {
			end = len(texts)
		}
		vecs, err := embedder.Embed(texts[i:end])
		if err != nil {
			return nil, err
		}
		result = append(result, vecs...)
	}
	return result, nil
}

// CosineSimilarity computes the cosine similarity between two vectors.
// Delegates to util.CosineSimilarity (canonical implementation).
func CosineSimilarity(a, b []float64) (float64, error) {
	return util.CosineSimilarity(a, b)
}

// cctToMemoryItem converts a CCT pattern to a MemoryItem for unified scoring.
// Uses SourceManual because no "synthesized" source exists; Context.Intent
// disambiguates these from genuinely manual entries.
func cctToMemoryItem(p compound.CctPattern) memory.MemoryItem {
	return memory.MemoryItem{
		ID:         p.ID,
		Type:       memory.TypeLesson,
		Trigger:    p.Name,
		Insight:    p.Description,
		Tags:       []string{},
		Source:      memory.SourceManual,
		Context:    memory.Context{Tool: "compound", Intent: "synthesis"},
		Created:    p.Created,
		Confirmed:  true,
		Supersedes: []string{},
		Related:    p.SourceIDs,
	}
}

// SearchVector performs vector similarity search over all items in the database
// and CCT patterns from cct-patterns.jsonl.
//
// Algorithm:
//  1. Read all non-invalidated items + CCT patterns.
//  2. Embed the query text.
//  3. For each item, use cached embedding if hash matches, otherwise embed and cache.
//  4. Compute cosine similarity, sort descending, return top `limit`.
func SearchVector(db *sql.DB, embedder Embedder, query string, limit int, repoRoot string) ([]ScoredItem, error) {
	sdb := storage.NewSearchDB(db)
	items, err := sdb.ReadAll()
	if err != nil {
		return nil, err
	}

	// Read CCT patterns if available
	var cctPatterns []compound.CctPattern
	if repoRoot != "" {
		cctPatterns, _ = compound.ReadCctPatterns(repoRoot)
	}

	if len(items) == 0 && len(cctPatterns) == 0 {
		return nil, nil
	}

	queryVecs, err := embedder.Embed([]string{query})
	if err != nil {
		return nil, err
	}
	queryVec := queryVecs[0]

	cache := storage.GetCachedEmbeddingsBulk(db)

	// Separate cached from uncached items for batched embedding
	type uncachedEntry struct {
		idx  int
		text string
		hash string
	}
	itemVecs := make([][]float64, len(items))
	var uncached []uncachedEntry

	for i, item := range items {
		hash := storage.ContentHash(item.Trigger, item.Insight)
		if cached, ok := cache[item.ID]; ok && cached.Hash == hash {
			itemVecs[i] = cached.Vector
		} else {
			uncached = append(uncached, uncachedEntry{idx: i, text: item.Trigger + " " + item.Insight, hash: hash})
		}
	}

	// Batch embed all uncached items (chunked to respect daemon limits)
	if len(uncached) > 0 {
		texts := make([]string, len(uncached))
		for i, u := range uncached {
			texts[i] = u.text
		}
		vecs, err := embedBatched(embedder, texts)
		if err != nil {
			return nil, err
		}
		for i, u := range uncached {
			itemVecs[u.idx] = vecs[i]
			_ = storage.SetCachedEmbedding(db, items[u.idx].ID, vecs[i], u.hash) // cache write failure is non-fatal; search proceeds with in-memory result
		}
	}

	var results []ScoredItem
	for i, item := range items {
		score, err := CosineSimilarity(queryVec, itemVecs[i])
		if err != nil {
			continue
		}
		results = append(results, ScoredItem{Item: item, Score: score})
	}

	// Score CCT patterns — batched with chunking for daemon limits
	if len(cctPatterns) > 0 {
		cctTexts := make([]string, len(cctPatterns))
		for i, p := range cctPatterns {
			cctTexts[i] = p.Name + " " + p.Description
		}
		cctVecs, cctErr := embedBatched(embedder, cctTexts)
		if cctErr == nil && len(cctVecs) == len(cctPatterns) {
			for i, pattern := range cctPatterns {
				score, err := CosineSimilarity(queryVec, cctVecs[i])
				if err != nil {
					continue
				}
				results = append(results, ScoredItem{Item: cctToMemoryItem(pattern), Score: score})
			}
		}
	}

	sort.Slice(results, func(i, j int) bool {
		return results[i].Score > results[j].Score
	})

	if limit > 0 && len(results) > limit {
		results = results[:limit]
	}

	return results, nil
}

// FindSimilarLessons finds items whose insight text is similar to the given text.
// Only items with similarity >= threshold are returned. The item with excludeID
// is skipped (useful to avoid matching an item against itself).
//
// Uses insight-only embeddings (not trigger+insight like SearchVector).
func FindSimilarLessons(db *sql.DB, embedder Embedder, text string, threshold float64, excludeID string) ([]ScoredItem, error) {
	sdb := storage.NewSearchDB(db)
	items, err := sdb.ReadAll()
	if err != nil {
		return nil, err
	}
	if len(items) == 0 {
		return nil, nil
	}

	queryVecs, err := embedder.Embed([]string{text})
	if err != nil {
		return nil, err
	}
	queryVec := queryVecs[0]

	cache := storage.GetCachedInsightEmbeddingsBulk(db)

	// Separate cached from uncached items for batched embedding
	type uncachedEntry struct {
		idx  int
		hash string
	}
	// Filter out excluded item and collect cache status
	type candidate struct {
		item memory.MemoryItem
		hash string
	}
	var candidates []candidate
	for _, item := range items {
		if item.ID == excludeID {
			continue
		}
		candidates = append(candidates, candidate{item: item, hash: storage.ContentHash(item.Insight, "")})
	}

	itemVecs := make([][]float64, len(candidates))
	var uncached []uncachedEntry

	for i, c := range candidates {
		if cached, ok := cache[c.item.ID]; ok && cached.Hash == c.hash {
			itemVecs[i] = cached.Vector
		} else {
			uncached = append(uncached, uncachedEntry{idx: i, hash: c.hash})
		}
	}

	// Batch embed all uncached items (chunked to respect daemon limits)
	if len(uncached) > 0 {
		texts := make([]string, len(uncached))
		for i, u := range uncached {
			texts[i] = candidates[u.idx].item.Insight
		}
		vecs, err := embedBatched(embedder, texts)
		if err != nil {
			return nil, err
		}
		for i, u := range uncached {
			itemVecs[u.idx] = vecs[i]
			_ = storage.SetCachedInsightEmbedding(db, candidates[u.idx].item.ID, vecs[i], u.hash) // cache write failure is non-fatal; search proceeds with in-memory result
		}
	}

	var results []ScoredItem
	for i, c := range candidates {
		score, err := CosineSimilarity(queryVec, itemVecs[i])
		if err != nil {
			continue
		}
		if score >= threshold {
			results = append(results, ScoredItem{Item: c.item, Score: score})
		}
	}

	sort.Slice(results, func(i, j int) bool {
		return results[i].Score > results[j].Score
	})

	return results, nil
}
